### ISSUE TITLE
Add YOLOv8 model pruning support (starting with PyTorch ln_structured)

### ISSUE THREAD

**Issue by syedhamzamohiuddin** [ultralytics/ultralytics#21587]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I'd like to contribute a pruning module for YOLOv8 starting with PyTorch's built-in ln_structured pruning (torch.nn.utils.prune.ln_structured).

My plan is to:

Provide a clean, modular pruning function that takes a YOLOv8 model and applies structured pruning to Conv layers.

Include a minimal example for loading, pruning, and saving the model.

Optionally include a simple script for fine-tuning after pruning.

This approach avoids external dependencies and keeps it simple.

Would you be open to this contribution?

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @syedhamzamohiuddin, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for general guidance, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as many common questions are already answered there.

This is an automated response to help you get started—an Ultralytics engineer will review your feature request and assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If your request is related to custom training or workflows, please include as much detail as possible, such as dataset samples, logs, or other relevant info, and check our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best! For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussions, check out [Discourse](https://community.ultralytics.com/). Or join threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share and learn from others.

## Upgrade

Before diving in, ensure you are using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of these up-to-date, verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your suggestion and willingness to contribute! An Ultralytics engineer will respond directly regarding model pruning support for YOLOv8 and your proposed approach.

**Comment by glenn-jocher** [comment]
Thanks for proposing this valuable contribution! Model pruning is indeed an important optimization technique for deploying YOLO models on edge devices, and using PyTorch's built-in `ln_structured` pruning is a clean approach.

We'd welcome this contribution - please follow our [contributing guidelines](https://docs.ultralytics.com/help/contributing/) to fork the repository, create a feature branch, and submit a PR with your implementation. Make sure to include tests and documentation for the new pruning functionality.

**Comment by syedhamzamohiuddin** [comment]
Thanks for the quick and positive response! I’ll proceed as suggested — I’ll experiment locally first to finalize the pruning implementation, and then fork the repo and open a PR from a feature branch. I’ll make sure to include documentation and tests for the functionality. Looking forward to contributing!

**Comment by glenn-jocher** [comment]
Excellent! Take your time with the local experimentation - that's a great approach to ensure the implementation is solid before submitting. Feel free to reach out in the issue if you run into any questions during development. We're looking forward to your contribution!

**Comment by syedhamzamohiuddin** [comment]
Sure, thank you! I’ll reach out if I have any questions during development.

**Comment by glenn-jocher** [comment]
Sounds good—when you open the PR, please mark it Draft for early feedback, keep pruning optional with no new deps, include a small round‑trip test (prune→save→load→inference) under tests, and the CLA bot will prompt you to sign.

**Comment by syedhamzamohiuddin** [comment]
Update: Implementation is now available in Draft PR #21977 for review.

**Comment by glenn-jocher** [comment]
Thanks—saw Draft PR #21977 and we’ll continue review there; please avoid adding config args to `default.yaml`, add a short docs page with a minimal usage example and current limitations, include a brief accuracy/speed note in the PR description, and ensure [GitHub Actions CI tests](https://docs.ultralytics.com/help/CI/) pass with unit coverage per our [tests guidance](https://docs.ultralytics.com/guides/model-testing/).
